go/mysql/datetime/interval.go

```diff
@@ -0,0 +1,425 @@
+	IntervalDaySecond         = IntervalDay | IntervalHour | IntervalMinute | IntervalSecond
+	IntervalDayMinute         = IntervalDay | IntervalHour | IntervalMinute
+	IntervalDayHour           = IntervalDay | IntervalHour
+	IntervalYearMonth         = IntervalYear | IntervalMonth
+)
+
+type intervalSetter func(tp *Interval, val int)
+
+var intervalSet = [...]intervalSetter{
+	intervalSetMicrosecond,
+	intervalSetSecond,
+	intervalSetMinute,
+	intervalSetHour,
+	intervalSetDay,
+	intervalSetMonth,
+	intervalSetYear,
+}
+
+// setter returns the setter method for this interval's type.
+// If this is a unary interval, it'll return the setter for the interval's unary type.
+// If this is a compound interval, it'll return the setter for the smallest unary type
+// in the interval.
+func (itv IntervalType) setter() intervalSetter {
+	// find the lowest bit set in the interval, this is the smallest unary type
+	unary := itv & -itv
+
+	// map from an unary interval type to its offset by counting the trailing
+	// zeroes. e.g. for HOUR(1 << 3), this will return 3, which the position
+	// for the HOUR setter in intervalSet
+	return intervalSet[bits.TrailingZeros8(uint8(unary))]
+}
+
+func (itv IntervalType) PartCount() int {
+	return bits.OnesCount8(uint8(itv & ^intervalMulti))
+}
+
+func (itv IntervalType) HasTimeParts() bool {
+	return itv&(IntervalHour|IntervalMinute|IntervalSecond|IntervalMicrosecond) != 0
+}
+
+func (itv IntervalType) HasDateParts() bool {
+	return itv&(IntervalYear|IntervalMonth|IntervalDay) != 0
+}
+
+func (itv IntervalType) HasDayParts() bool {
+	return (itv & IntervalDay) != 0
+}
+
+func (itv IntervalType) HasMonthParts() bool {
+	return (itv & IntervalMonth) != 0
+}
+
+func (itv IntervalType) NeedsPrecision() bool {
+	return itv&IntervalMicrosecond != 0
+}
+
+// ToString returns the type as a string
+func (itv IntervalType) ToString() string {
+	switch itv {
+	case IntervalYear:
+		return "year"
+	case IntervalQuarter:
+		return "quarter"
+	case IntervalMonth:
+		return "month"
+	case IntervalWeek:
+		return "week"
+	case IntervalDay:
+		return "day"
+	case IntervalHour:
+		return "hour"
+	case IntervalMinute:
+		return "minute"
+	case IntervalSecond:
+		return "second"
+	case IntervalMicrosecond:
+		return "microsecond"
+	case IntervalYearMonth:
+		return "year_month"
+	case IntervalDayHour:
+		return "day_hour"
+	case IntervalDayMinute:
+		return "day_minute"
+	case IntervalDaySecond:
+		return "day_second"
+	case IntervalHourMinute:
+		return "hour_minute"
+	case IntervalHourSecond:
+		return "hour_second"
+	case IntervalMinuteSecond:
+		return "minute_second"
+	case IntervalDayMicrosecond:
+		return "day_microsecond"
+	case IntervalHourMicrosecond:
+		return "hour_microsecond"
+	case IntervalMinuteMicrosecond:
+		return "minute_microsecond"
+	case IntervalSecondMicrosecond:
+		return "second_microsecond"
+	default:
+		return "[unknown IntervalType]"
+	}
+}
+
+func intervalSetYear(tp *Interval, val int) {
```

Comment:
nitpick: why is this and the following methods not methods with a `*Interval` receiver?